Framework/Indexing/inc/MantidIndexing/SpectrumNumberTranslator.h

```diff
@@ -0,0 +1,154 @@
+#ifndef MANTID_INDEXING_SPECTRUMNUMBERTRANSLATOR_H_
+#define MANTID_INDEXING_SPECTRUMNUMBERTRANSLATOR_H_
+
+#include "MantidIndexing/DllConfig.h"
+#include "MantidIndexing/Partitioning.h"
+#include "MantidIndexing/SpectrumIndexSet.h"
+#include "MantidIndexing/SpectrumNumber.h"
+
+#include <algorithm>
+#include <unordered_map>
+
+namespace Mantid {
+namespace Indexing {
+
+/** SpectrumNumberTranslator : TODO: DESCRIPTION
+
+  @author Simon Heybrock
+  @date 2016
+
+  Copyright &copy; 2016 ISIS Rutherford Appleton Laboratory, NScD Oak Ridge
+  National Laboratory & European Spallation Source
+
+  This file is part of Mantid.
+
+  Mantid is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 3 of the License, or
+  (at your option) any later version.
+
+  Mantid is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+  File change history is stored at: <https://github.com/mantidproject/mantid>
+  Code Documentation is available at: <http://doxygen.mantidproject.org>
+*/
+class MANTID_INDEXING_DLL SpectrumNumberTranslator {
+public:
+  SpectrumNumberTranslator(std::vector<SpectrumNumber> spectrumNumbers,
+                           const Partitioning &partitioning,
+                           const PartitionIndex &partition)
+      : m_partition(partition) {
+    partitioning.checkValid(m_partition);
+
+    // This sort is the reason for taking the spectrumNumbers argument by value.
+    // We must sort such that the min/max variant of makeIndexSet() makes sense.
+    std::sort(spectrumNumbers.begin(), spectrumNumbers.end());
+
+    size_t currentIndex = 0;
+    bool first = true;
+    for (size_t i = 0; i < spectrumNumbers.size(); ++i) {
+      auto number = spectrumNumbers[i];
+      auto partition = partitioning.indexOf(number);
+      m_partitions.emplace(number, partition);
+      if (partition == m_partition) {
+        m_indices.emplace(number, currentIndex++);
+        // Store min spectrum number in this partition
+        if (first) {
+          m_min = number;
+          first = false;
+        }
+        // Store max spectrum number in this partition
+        m_max = number;
+      }
+    }
+    if (spectrumNumbers.size() != m_partitions.size())
+      throw std::logic_error("SpectrumNumberTranslator: The vector of spectrum "
+                             "numbers contained duplicate entries.");
+  }
+
+  // Full set
+  SpectrumIndexSet makeIndexSet() { return SpectrumIndexSet(m_indices.size()); }
+
+  SpectrumIndexSet makeIndexSet(SpectrumNumber min, SpectrumNumber max) {
```

Comment:
@OwenArnold The implementation of this min/max variant is really nasty/ugly. It would be much simpler if we would use a `std::map` instead of `std::unordered_map`, but I am not sure if that would cause trouble with performance. Do you have a feeling of how frequently different translation requests would occur?